ReactiveCocoa/Swift/Lifetime.swift

```diff
@@ -0,0 +1,36 @@
+import enum Result.NoError
+
+/// Provides a signal that completes when the lifetime object is deinitialized.
+///
+/// When assigned to a property of another object, provides a hook to
+/// observe when that object goes out of scope.
+public final class Lifetime {
+	/// A signal that sends a Completed event when the lifetime ends.
+	public let ended: Signal<(), NoError>
+
+	private let endedObserver: Observer<(), NoError>
+
+	public init() {
+		(ended, endedObserver) = Signal.pipe()
+	}
+
+	deinit {
+		endedObserver.sendCompleted()
+	}
+}
+
+private var lifetimeKey: UInt8 = 0
+
+extension NSObject {
+	/// Returns a lifetime that ends when the receiver is deallocated.
+	@nonobjc public var lifetime: Lifetime {
+		if let lifetime = objc_getAssociatedObject(self, &lifetimeKey) as! Lifetime? {
```

Comment:
Yep, that's a good call. I've pushed a change which synchronises on `self` here.

> P.S. Why not simply write as? Lifetime?

I originally used that, see [this discussion](https://github.com/ReactiveCocoa/ReactiveCocoa/pull/3030#discussion_r69509093). There is no valid reason for the type of the associated object to be something other than `Lifetime`. If it is something else, then it should be caught as early as possible in development.